### PR TITLE
[ENG-1475] Last login optimization

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Update last login field to go faster!

## Changes

- change `date_last_login` to db index field
- isolate updating code, remove seemly unnecessary join
- add tests

## QA Notes

- No testable migrations
- Low risk
- No permission code touched
- no added features
- QA can verify by just logging and out of the site
- This should be slightly faster

## Documentation

optimization, no new docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1475

### CODE REVIEW

File: pytest.ini [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts = --ds=osf_tests.settings --tb=short --reuse-db --allow-hosts=127.0.0.1
+addopts = --ds=osf_tests.settings --tb=short --reuse-db --allow-hosts=127.0.0.1,192.168.168.167
```
Comment (Johnetordoff) [inline pytest.ini]: This I'm not quite sure about, but I have to include all the time when debugging, it allows the test app to make requests to it's own container.

File: framework/sessions/__init__.py [hunk 2]
```diff
@@ -181,3 +174,13 @@ def after_request(response):
...
     return response
+
+
+def update_user_date_last_login(user_session):
+    OSFUser = apps.get_model('osf.OSFUser')
+    now = timezone.now()
+    OSFUser.objects.filter(
+        Q(date_last_login__isnull=True) |
+        Q(date_last_login__lt=now - settings.DATE_LAST_LOGIN_THROTTLE_DELTA),
+        username=user_session.data['auth_user_username'],
```
Comment (Johnetordoff) [inline framework/sessions/__init__.py]: No more join with the guid table, saves some time I believe!

File: framework/sessions/__init__.py [hunk 3]
```diff
@@ -165,14 +164,8 @@ def before_request():
...
-                OSFUser = apps.get_model('osf.OSFUser')
-                (
-                    OSFUser.objects
-                    .filter(guids___id__isnull=False, guids___id=user_session.data['auth_user_id'])
-                    # Throttle updates
-                    .filter(Q(date_last_login__isnull=True) | Q(date_last_login__lt=timezone.now() - dt.timedelta(seconds=settings.DATE_LAST_LOGIN_THROTTLE)))
-                ).update(date_last_login=timezone.now())
-            set_session(user_session)
+                update_user_date_last_login(user_session)
+                set_session(user_session)
```
Comment (corbinSanders) [inline framework/sessions/__init__.py]: Did you mean to move `set_session(user_session)` into the `if user_session.data...` if block?

Comment (Johnetordoff) [reply framework/sessions/__init__.py]: Good catch!

File: osf_tests/users/test_last_login_date.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    def test_date_last_login_updated(self, mock_time):
+        now = datetime(2018, 2, 4, tzinfo=pytz.utc)
+        mock_time.return_value = now
+        assert self.user.date_last_login is None
+
+        self.app.set_cookie(settings.COOKIE_NAME, self.cookie)
+        self.app.get(f'{settings.DOMAIN}{self.user._id}')  # user page will fail because not emberized
+
+        self.user.refresh_from_db()
+        assert self.user.date_last_login == now
```
Comment (brianjgeiger) [inline osf_tests/users/test_last_login_date.py]: Should there be three test cases: `date_last_login` is `None`, `date_last_login` is less than the threshold, and `date_last_login` is greater than the threshold?

**Comment by Johnetordoff** [comment]
This passes Travis on my branch https://travis-ci.org/Johnetordoff/osf.io/builds/652122967?utm_medium=notification&utm_source=github_status

**Pull request merged**